ext/lib/crypto/mbedtls/Kconfig.tls-generic

```diff
@@ -0,0 +1,217 @@
+	default n
+
+config TLS_KEY_EXCHANGE_ECJPAKE_ENABLED
+	bool "Enable the ECJPAKE based ciphersuite modes"
+	default n
+
+if TLS_KEY_EXCHANGE_ECDHE_PSK_ENABLED || \
+   TLS_KEY_EXCHANGE_ECDHE_RSA_ENABLED || \
+   TLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED || \
+   TLS_KEY_EXCHANGE_ECDH_ECDSA_ENABLED || \
+   TLS_KEY_EXCHANGE_ECDH_RSA_ENABLED || \
+   TLS_KEY_EXCHANGE_ECJPAKE_ENABLED \
+
+comment "Supported elliptic curves"
+
+config TLS_ECP_DP_SECP192R1_ENABLED
+	bool "Enable SECP192R1 elliptic curve"
+	default n
+
+config TLS_ECP_DP_SECP224R1_ENABLED
+	bool "Enable SECP224R1 elliptic curve"
+	default n
+
+config TLS_ECP_DP_SECP256R1_ENABLED
+	bool "Enable SECP256R1 elliptic curve"
+	default n
+
+config TLS_ECP_DP_SECP384R1_ENABLED
+	bool "Enable SECP384R1 elliptic curve"
+	default n
+
+config TLS_ECP_DP_SECP521R1_ENABLED
+	bool "Enable SECP521R1 elliptic curve"
+	default n
+
+config TLS_ECP_DP_SECP192K1_ENABLED
+	bool "Enable SECP192K1 elliptic curve"
+	default n
+
+config TLS_ECP_DP_SECP224K1_ENABLED
+	bool "Enable SECP224K1 elliptic curve"
+	default n
+
+config TLS_ECP_DP_SECP256K1_ENABLED
+	bool "Enable SECP256K1 elliptic curve"
+	default n
+
+config TLS_ECP_DP_BP256R1_ENABLED
+	bool "Enable BP256R1 elliptic curve"
+	default n
+
+config TLS_ECP_DP_BP384R1_ENABLED
+	bool "Enable BP384R1 elliptic curve"
+	default n
+
+config TLS_ECP_DP_BP512R1_ENABLED
+	bool "Enable BP512R1 elliptic curve"
+	default n
+
+config TLS_ECP_DP_CURVE25519_ENABLED
+	bool "Enable CURVE25519 elliptic curve"
+	default n
+
+config TLS_ECP_DP_CURVE448_ENABLED
+	bool "Enable CURVE448 elliptic curve"
+	default n
+
+config TLS_ECP_NIST_OPTIM
+	bool "Enable NSIT curves optimization"
+	default n
+
+endif
+
+comment "Supported cipher modes"
+
+config TLS_CIPHER_AES_ENABLED
+	bool "Enable the AES block cipher"
+	default y
+
+config TLS_CIPHER_CAMELLIA_ENABLED
+	bool "Enable the Camellia block cipher"
+	default n
+
+config TLS_CIPHER_DES_ENABLED
+	bool "Enable the DES block cipher"
+	default y
+
+config TLS_CIPHER_CCM_ENABLED
+	bool "Enable the Counter with CBC-MAC (CCM) mode for 128-bit block cipher"
+	default n
+	depends on TLS_CIPHER_AES_ENABLED || TLS_CIPHER_CAMELLIA_ENABLED
+
+config TLS_CIPHER_CBC_ENABLED
+	bool "Enable Cipher Block Chaining mode (CBC) for symmetric ciphers"
+	default y
+
+comment "Supported message authentication methods"
+
+config TLS_MAC_MD5_ENABLED
+	bool "Enable the MD5 hash algorithm"
+	default y
+
+config TLS_MAC_SHA1_ENABLED
+	bool "Enable the SHA1 hash algorithm"
+	default y
+
+config TLS_MAC_SHA256_ENABLED
+	bool "Enable the SHA-224 and SHA-256 hash algorithms"
+	default y
+
+config TLS_MAC_SHA512_ENABLED
+	bool "Enable the SHA-384 and SHA-512 hash algorithms"
+	default n
+
+endmenu
+
+config TLS_PEM_CERTIFICATE_FORMAT
+	bool "Enable support for PEM certificate format"
+	default n
+	help
+	  By default only DER (binary) format of certificates is supported. Enable
+	  this option to enable support for PEM format.
+
+config TLS_USER_CONFIG_ENABLE
```

Comment:
Did you consider having just a single Kconfig options for this? E.g. if  TLS_USER_CONFIG_FILE is defined to non-empty value, #include it?

Reply:
Yes, that was my initial thought, but that would require to check for an empty string in preprocessor, and I didn't find an easy way to do that. Any suggestions on this?

Reply:
Yeah, I never thought what `#if ""` would evaluate too, but it's actually an error to use strings in #if at all:

~~~
a.c:1:11: error: token ""foo"" is not valid in preprocessor expressions
 #define a "foo"
           ^
a.c:3:5: note: in expansion of macro ‘a’
 #if a
     ^
~~~

So, nevermind.